lib/ansible/module_utils/api.py

```diff
@@ -91,26 +92,58 @@ def ratelimited(*args, **kwargs):
     return wrapper
 
 
-def retry(retries=None, retry_pause=1):
-    """Retry decorator"""
-    def wrapper(f):
+def _retry_limit_exceeded(retries):
+    def function_wrapper(function):
+        def raise_error(*args, **kwargs):
+            raise Exception("Retry limit exceeded: %d" % retries)
+        return raise_error
+    return function_wrapper
 
-        def retried(*args, **kwargs):
-            retry_count = 0
-            if retries is not None:
-                ret = None
-                while True:
-                    retry_count += 1
-                    if retry_count >= retries:
-                        raise Exception("Retry limit exceeded: %d" % retries)
-                    try:
-                        ret = f(*args, **kwargs)
-                    except Exception:
-                        pass
-                    if ret:
-                        break
-                    time.sleep(retry_pause)
-                return ret
-
-        return retried
-    return wrapper
+
+def retry(retries=None, retry_pause=1):
+    """Retry decorator.
+
+    This assumes the function has already been called, so if retries == 1, the maximum attempts
+    have been reached. If retries is None, the function is a no-op.
+    """
+    if retries is None:
+        return lambda function: lambda *args, **kwargs: None
+    if retries <= 1:
+        return _retry_limit_exceeded(retries)
+    try:
+        backoff_delays = [retry_pause for delay in range(0, retries)]
+        return retry_wrapper(backoff_iterator=backoff_delays, retry_condition=lambda x: True)
+    except Exception:
+        raise Exception("Retry limit exceeded: %d" % retries)
+
+
+def jittered_backoff_generator(retries=10, delay=3, max_delay=60):
+    for retry in range(0, retries):
+        yield random.randint(0, min(max_delay, delay * 2 ** retry))
+
+
+def retry_wrapper(backoff_iterator, retry_condition):
```

Comment:
```suggestion
def retry_with_delays(backoff_iterator, retry_condition):
```
(wrapper is an implementation detail, it's probably more informative to say that it adds some delays)

Also, thinking about `backoff_iterator` — maybe it's more like a strategy? Not sure tho.

Reply:
backoff_iterator and strategy make me think of a delay generator, not a decorator doing the retries

`retry_with_delays_and_condition`?

Reply:
Yeah, probably. It feels like you're getting closer to implementing the API of https://github.com/litl/backoff — so maybe just copy their invocation signature (at least the args that make sense for us) and go from there? It's pretty neat.

Reply:
I don't see the point to incorporating a 3rd party lib unnecessarily. I do think it's neat, but this was loosely based on the old lib/ansible/module_utils/ec2.py stuff. That stuff has worked for ages with a wide variety of AWS services all off doing their own thing, why reinvent the wheel.

Reply:
I'm not saying that we should use `backoff`, I say that we should adopt some of its API definitions (as in the types of arguments `@backoff()` accepts — some, not all).

Reply:
Can you give me a link with the signature you're thinking of? I don't really know what you're referring to in particular, and I don't see it from a quick scan of the repo.

Reply:
I was thinking along the lines of https://github.com/litl/backoff/blob/e20982dc36be213465a1e26864d8e9792fd3af92/backoff/_decorator.py#L18 / https://github.com/litl/backoff/blob/e20982dc36be213465a1e26864d8e9792fd3af92/backoff/_decorator.py#L105.

I think the useful interface parts for us are:
```python
@backoff.on_predicate(
    wait_gen,  # iterator, emitter of delays
    predicate,  # a callable accepting the decorated function return value and deciding whether to retry on not (returns bool)
    max_tries,
    max_time,
    jitter,  # delay transformer callable, converts values produced by wait_gen
)
```
and
```python
@backoff.on_exception(
    wait_gen,  # iterator, emitter of delays
    exception,  # exception triggering retry
    max_tries,
    max_time,
    jitter,  # delay transformer callable, converts values produced by wait_gen
    giveup,  # optional callable analyzing the exception additionally to the `exception` argument, can judge whether to retry or reraise
)
```

I like that this is composed of decoupled components — `wait_gen` and `giveup` are separate rather than coupled, retries/time are not tied to jitter and wait_gen. And the way to cleanly provide a retriable exception (could be a tuple I guess) is just brilliant — it allows the reader of the calling code to easily reason about what it does.

The helper names could be `retry_on_exception` and `retry_on_if`. Although, maybe we only need one variant of this interface, not both.

Reply:
* wait_gen
* max_tries
* max_time
* jitter

Taking these seems less generic than taking a generator that yields delays because it's implying an implementation. For example, if this were to be used in the `reboot` module, the delays would be yielded until a max total duration (which seems like a nice thing to support from a user-perspective), not a pre-determined number of attempts.

* predicate

This is nice, basically you can have a retry decorator double as a waiter. There's nothing that would use it yet, but maybe I should add it as an argument now so we don't need to add it as a kwarg later for backwards compatibility? It will make it easy to handle the existing galaxy/api.py retries when that's refactored.

* exception

What's the benefit to taking a specific classes when the conditional can check `isinstance`?

* giveup

The current name seems fitting since this is operating on the assumption that no exceptions will be retried by default instead of backoff's assumption that all will be retried.

Reply:
I added a 'retry_on_success' parameter similar to predicate. Let me know what you think - I know the signature is not the same but functionality-wise, this feels a little more flexible to me.

Reply:
> What's the benefit to taking a specific classes when the conditional can check `isinstance`?

Just for sugar but yeah, that's not mandatory.

> I added a 'retry_on_success' parameter similar to predicate. Let me know what you think - I know the signature is not the same but functionality-wise, this feels a little more flexible to me.

Yep, the signature doesn't need to be exactly the same. I just thought it'd demonstrate better what I meant.

I feel like `retry_on_success` reads like it accepts a boolean (rather than a callable) that will cause a retry whenever the decorated $thing succeeds. So I'd probably choose a different name.